transcode/transcode.go

```diff
@@ -115,59 +115,40 @@ func RunTranscodeProcess(transcodeRequest TranscodeSegmentRequest, streamName st
 	for segmentIndex, u := range sourceSegmentURLs {
-		rc, err := clients.DownloadOSURL(u.URL)
-		if err != nil {
-			return outputs, fmt.Errorf("failed to download source segment %q: %s", u, err)
-		}
-
-		var tr clients.TranscodeResult
-		// If an AccessToken is provided via the request for transcode, then use remote Broadcasters.
-		// Otherwise, use the local harcoded Broadcaster.
-		if transcodeRequest.AccessToken != "" {
-			creds := clients.Credentials{
-				AccessToken:  transcodeRequest.AccessToken,
-				CustomAPIURL: transcodeRequest.TranscodeAPIUrl,
-			}
-			broadcasterClient, _ := clients.NewRemoteBroadcasterClient(creds)
-			// Get renditions from remote broadcaster
-			tr, err = broadcasterClient.TranscodeSegmentWithRemoteBroadcaster(rc, int64(segmentIndex), transcodeProfiles, streamName, u.DurationMillis)
-			if err != nil {
-				return outputs, fmt.Errorf("failed to run TranscodeSegmentWithRemoteBroadcaster: %s", err)
-			}
-		} else {
-			// Get renditions from local broadcaster
-			tr, err = localBroadcasterClient.TranscodeSegment(rc, int64(segmentIndex), transcodeProfiles, u.DurationMillis, manifestID)
-			if err != nil {
-				return outputs, fmt.Errorf("failed to run TranscodeSegment: %s", err)
-			}
-		}
-		// Store renditions
-		for _, transcodedSegment := range tr.Renditions {
-			renditionIndex := getProfileIndex(transcodeProfiles, transcodedSegment.Name)
-			if renditionIndex == -1 {
-				return outputs, fmt.Errorf("failed to find profile with name %q while parsing rendition segment", transcodedSegment.Name)
-			}
-
-			relativeRenditionPath := fmt.Sprintf("rendition-%d/", renditionIndex)
-			relativeRenditionURL, err := url.Parse(relativeRenditionPath)
-			if err != nil {
-				return outputs, fmt.Errorf("error building rendition segment URL %q: %s", relativeRenditionPath, err)
-			}
-			renditionURL := targetOSURL.ResolveReference(relativeRenditionURL)
-
-			err = clients.UploadToOSURL(renditionURL.String(), fmt.Sprintf("%d.ts", segmentIndex), bytes.NewReader(transcodedSegment.MediaData))
-			if err != nil {
-				return outputs, fmt.Errorf("failed to upload master playlist: %s", err)
+		queue <- segmentInfo{Input: u, Index: segmentIndex}
+	}
+	close(queue)
+	// Use channel for recieving the errors
+	errors := make(chan error, 100)
+	// Start number of workers in parallel
+	var completed sync.WaitGroup
+	completed.Add(config.TranscodingParallelJobs)
+	for index := 0; index < config.TranscodingParallelJobs; index++ {
+		go func() {
+			defer completed.Done()
+			for segment := range queue {
+				err := transcodeSegment(segment, streamName, manifestID, transcodeRequest, transcodeProfiles, targetOSURL)
+				if err != nil {
+					errors <- err
+					return
+				}
+				var completedRatio = calculateCompletedRatio(len(sourceSegmentURLs), segment.Index+1)
+				if err = clients.DefaultCallbackClient.SendTranscodeStatus(transcodeRequest.CallbackURL, clients.TranscodeStatusTranscoding, completedRatio); err != nil {
```

Comment:
hmm won't we get multiple completed callbacks now instead of just the one that Studio was expecting?

Reply:
This is actually progress callback that looks like:
```json
{
  "completion_ratio": 0.6666666666666667,
  "status": "transcoding",
  "timestamp": 1667312537
}
```

And final completed callback is [here](https://github.com/livepeer/catalyst-api/blob/ak/parallel_seg_push/transcode/transcode.go#L167):
```json
{
  "completion_ratio": 1,
  "status": "success",
  "timestamp": 1667317010,
  "type": "video",
  "video_spec": {
    "format": "mp4",
    "tracks": [
      {
        "type": "video",
        "codec": "H264",
        "bitrate": 7508656,
        "duration": 184.541,
        "size": 0,
        "start_time": 0,
        "width": 2048,
        "height": 858
      },
      {
        "type": "audio",
        "codec": "AAC",
        "bitrate": 256000,
        "duration": 184.661,
        "size": 0,
        "start_time": 0,
        "channels": 2,
        "sample_rate": 48000,
        "sample_bits": 16
      }
    ],
    "duration": 184.64,
    "size": 183755148
  },
  "outputs": [
    {
      "type": "google-s3",
      "manifest": "s3+https:/GOOG1EXCXWNKVLKBEJS62OMGXMQ5C5FH67YZMQXYGFBQYQRBYXQQJT5OM6L4I:rzM9CFtiunoflGzBwU3c%2FOBw9D5hy0Ntbmmedu6x@storage.googleapis.com/alexk-dms-upload-test/bbb/long/transcoded/index.m3u8",
      "videos": null
    }
  ]
}
```